src/rust/graph-query/src/lib.rs

```diff
@@ -0,0 +1,626 @@
+            // filter_string.push(')');
+            if i < self.property_filters.len() - 1 {
+                filter_string.push_str(" AND ");
+            }
+        }
+
+        if self.property_filters.len() >= 1 {
+            filter_string.push(')');
+        }
+    }
+
+    fn to_query(
+        &self,
+        binding: u8,
+        var_allocator: &mut VarAllocator,
+        query_string: &mut String,
+        visited: &mut HashSet<(u128, String, u128)>,
+        should_filter: bool,
+    ) {
+        if should_filter {
+            self.to_filter(query_string, var_allocator);
+        }
+
+        query_string.push('{');
+        for property_name in self.property_filters.keys() {
+            query_string.push('\n');
+            query_string.push_str(property_name);
+        }
+        for (edge_name, edges) in self.edge_filters.iter() {
+            for edge in edges {
+                let inner_edge = edge.0.clone();
+                let inner_edge: &RefCell<NodeQuery> = inner_edge.borrow();
+                let inner_edge = inner_edge.borrow();
+
+                if self.already_visited(edge_name, edge.get_id(), visited) {
+                    continue;
+                }
+
+                query_string.push('\n');
+                if edge.is_root() {
+                    query_string.push_str("binding_");
+                    query_string.push_str(&binding.to_string());
+                    query_string.push_str(" as ");
+                }
+                query_string.push_str(edge_name);
+                inner_edge.to_query(binding, var_allocator, query_string, visited, should_filter);
+            }
+        }
+        query_string.push('}');
+    }
+
+    // When building our query we need to understand if we've visited a node either through a forward
+    // or a reverse edge
+    // We can't just check if we have visited the destination node because multiple nodes may reference
+    // another node, or a single node may reference another node with different edge names
+    // For example, a Process may both Read and Write to a File
+    fn already_visited(
+        &self,
+        edge_name: &str,
+        edge_id: u128,
+        visited: &mut HashSet<(u128, String, u128)>,
+    ) -> bool {
+        // Rust's hashmap is kinda annoying about borrowing keys that have nested owned values. Feel free
+        // to remove this unnecessary allocation if you're willing to put the work in (I am not).
+        // https://users.rust-lang.org/t/hashmap-with-tuple-keys/12711
+        let f_key = (self.id, edge_name.to_owned(), edge_id);
+        if visited.contains(&f_key) {
+            return true;
+        }
+        visited.insert(f_key);
+
+        let r_key = (
+            edge_id,
+            self.reverse_edge_names[edge_name].to_owned(),
+            self.id,
+        );
+        if visited.contains(&r_key) {
+            return true;
+        }
+        visited.insert(r_key);
+
+        false
+    }
+}
```

Comment:
I don't really see the point tbh. The functions are all internal, they produce strings, which are very hard to validate manually, and we already validate all of the string generation by talking to dgraph and having it parse the query.

Reply:
I don't see how strings are hard to validate... in my experience, they're some of the _easiset_ things to validate with a test.

Reply:
How would you validate the string:
`(NOT eq($a, "bar"))`
?
Also knowing that it could be generated as:
`NOT eq($a, "bar")`
or
`((NOT eq($a, "bar")))`
or have spaces added, or removed.

I could search for "NOT", and "$" and "eq" and "bar" ?

To me, the validation of a string is done by parsing the query. The parser exists in the dgraph database and occurs when we send that query to dgraph.